packages/compiler/src/experimental/mutators.ts

```diff
@@ -0,0 +1,389 @@
+}
+
+export interface Mutator {
+  name: string;
+  Model?: MutatorRecord<Model>;
+  ModelProperty?: MutatorRecord<ModelProperty>;
+  Scalar?: MutatorRecord<Scalar>;
+  Enum?: MutatorRecord<Enum>;
+  EnumMember?: MutatorRecord<EnumMember>;
+  Union?: MutatorRecord<Union>;
+  UnionVariant?: MutatorRecord<UnionVariant>;
+  Tuple?: MutatorRecord<Tuple>;
+  Operation?: MutatorRecord<Operation>;
+  Interface?: MutatorRecord<Interface>;
+  String?: MutatorRecord<Scalar>;
+  Number?: MutatorRecord<Scalar>;
+  Boolean?: MutatorRecord<Scalar>;
+  ScalarConstructor?: MutatorRecord<ScalarConstructor>;
+  StringTemplate?: MutatorRecord<StringTemplate>;
+  StringTemplateSpan?: MutatorRecord<StringTemplateSpan>;
+}
+
+export interface VisibilityOptions {
+  visibility: string;
+}
+
+export enum MutatorFlow {
+  MutateAndRecurse = 0,
+  DoNotMutate = 1 << 0,
+  DoNotRecurse = 1 << 1,
+}
+
+export function createVisibilityMutator(visibility: string): Mutator {
+  return {
+    name: visibility + " Visibility",
+    Model: {
+      filter(m, program, realm) {
+        if (isArrayModelType(program, m)) {
+          return MutatorFlow.DoNotMutate;
+        }
+        return true;
+      },
+      mutate(m, clone, program, realm) {
+        if (clone.name) {
+          clone.name = m.name + visibility.charAt(0).toUpperCase() + visibility.slice(1);
+        }
+
+        for (const prop of m.properties.values()) {
+          if (!isVisible(program, prop, [visibility])) {
+            clone.properties.delete(prop.name);
+            realm.remove(prop);
+          }
+        }
+        return;
+      },
+    },
+  };
+}
+
+const JSONMergePatch: Mutator = {
+  name: "JSON Merge Patch",
+  Model: {
+    filter(m, program, realm) {
+      // TODO: Revisit this, as it is not ideal
+      if (m.node!.parent!.kind === SyntaxKind.OperationSignatureDeclaration) {
+        return MutatorFlow.DoNotMutate;
+      }
+
+      return isArrayModelType(program, m)
+        ? MutatorFlow.DoNotRecurse | MutatorFlow.DoNotMutate
+        : true;
+    },
+    mutate(sourceType, clone, program, realm) {
+      if (clone.name) {
+        clone.name = clone.name + "MergePatch";
+      }
+      const typeFactory = createTypeFactory(program, realm);
+
+      for (const prop of clone.properties.values()) {
+        const clonedProp = realm.typeFactory.initializeClone(prop);
+        if (clonedProp.optional) {
+          if (clonedProp.type.kind === "Scalar") {
+            // remove everything but doc and apply it to the declaration
+            // TODO: THIS IS A HACK
+            const docDecorator = clonedProp.decorators.filter((d) => d.decorator === $doc);
+            const otherDecorators: [DecoratorFunction, ...any][] = clonedProp.decorators
+              .filter((d) => d.decorator !== $doc)
+              .map((d) => {
+                return [d.decorator, ...d.args.map((v) => v.jsValue)];
+              });
+
+            clonedProp.decorators = docDecorator;
```

Comment:
is this correct? and why does doc decorator needs special treatmeant might be good to add that to the hack comment above

Reply:
Yeah I think I'll remove these mutators for now until we have a better idea on how/where/if we want to expose those specific implementations.